packages/dataviews/src/view-list.js

```diff
@@ -67,66 +70,89 @@ export default function ViewList( {
 	}
 
 	return (
-		<ul className="dataviews-view-list">
-			{ usedData.map( ( item ) => {
+		<TreeGrid className="dataviews-view-list">
+			{ usedData.map( ( item, index ) => {
 				return (
-					<li
+					<TreeGridRow
 						key={ getItemId( item ) }
+						level={ 1 }
+						setSize={ 1 }
+						positionInSet={ index + 1 }
 						className={ classNames( {
 							'is-selected': selection.includes( item.id ),
 						} ) }
 					>
-						<HStack className="dataviews-view-list__item-wrapper">
-							<div
-								role="button"
-								tabIndex={ 0 }
-								aria-pressed={ selection.includes( item.id ) }
-								onKeyDown={ onEnter( item ) }
-								className="dataviews-view-list__item"
-								onClick={ () => onSelectionChange( [ item ] ) }
-							>
-								<HStack spacing={ 3 } justify="start">
-									<div className="dataviews-view-list__media-wrapper">
-										{ mediaField?.render( { item } ) || (
-											<div className="dataviews-view-list__media-placeholder"></div>
-										) }
-									</div>
-									<VStack spacing={ 1 }>
-										<span className="dataviews-view-list__primary-field">
-											{ primaryField?.render( { item } ) }
-										</span>
-										<div className="dataviews-view-list__fields">
-											{ visibleFields.map( ( field ) => {
-												return (
-													<span
-														key={ field.id }
-														className="dataviews-view-list__field"
-													>
-														{ field.render( {
-															item,
-														} ) }
-													</span>
-												);
-											} ) }
-										</div>
-									</VStack>
-								</HStack>
-							</div>
-							{ onDetailsChange && (
-								<Button
-									className="dataviews-view-list__details-button"
+						<TreeGridCell>
```

Comment:
I'm trying to use the `TreeGrid` component in `DataViews`. In `trunk`, using the list layout, the user moves with tabs through the list of items. See:

https://github.com/WordPress/gutenberg/assets/583546/689ce10e-2025-4482-9d2a-00f0c2c2937e

Note how this `div` is acting as a button, and it's a tab stop. When migrating to use `TreeGrid`, this element is skipped: the focus goes to the next `TreeGridCell` (details button) instead.

https://github.com/WordPress/gutenberg/assets/583546/c9496fbb-5182-462a-a7e2-09948b3e1157

Once the focus is in the button, I cannot reach this div through the arrow keys either (right/left):

https://github.com/WordPress/gutenberg/assets/583546/209d40ba-99ba-4b8a-8fac-3bc5afe42623

Reply:
Pinging people I've seen contributed to this and/or may be interested in following-up. @mirka @andrewhayward @ciampo @talldan @tellthemachines @andrewserong

Reply:
I've tried the following: setting the tabIndex to 0 when the focus is outside the list view and let `TreeGrid` handle it when it's inside. It doesn't work, once the focus is moved via the arrow keys to the next item, the details button becomes focused:

https://github.com/WordPress/gutenberg/assets/583546/4cec26e5-e3ec-4ddd-a2ac-a71b015cf909

<details>
<summary>Diff in case anyone wants to try</summary>

```diff
diff --git a/packages/dataviews/src/view-list.js b/packages/dataviews/src/view-list.js
index ce1f3a2b4f..18949f2bd5 100644
--- a/packages/dataviews/src/view-list.js
+++ b/packages/dataviews/src/view-list.js
@@ -18,6 +18,7 @@ import {
 import { ENTER, SPACE } from '@wordpress/keycodes';
 import { info } from '@wordpress/icons';
 import { __ } from '@wordpress/i18n';
+import { useRef } from '@wordpress/element';
 
 export default function ViewList( {
        view,
@@ -30,6 +31,7 @@ export default function ViewList( {
        selection,
        deferredRendering,
 } ) {
+       const treeGridRef = useRef();
        const shownData = useAsyncList( data, { step: 3 } );
        const usedData = deferredRendering ? shownData : data;
        const mediaField = fields.find(
@@ -69,8 +71,15 @@ export default function ViewList( {
                );
        }
 
+       const isFocusWithin = () => {
+               if ( ! treeGridRef.current ) {
+                       return false;
+               }
+               return treeGridRef.current.contains( document.activeElement );
+       }
+
        return (
-               <TreeGrid className="dataviews-view-list">
+               <TreeGrid ref={ treeGridRef } className="dataviews-view-list">
                        { usedData.map( ( item, index ) => {
                                return (
                                        <TreeGridRow
@@ -95,7 +104,7 @@ export default function ViewList( {
                                                                                onSelectionChange( [ item ] )
                                                                        }
                                                                        { ...otherProps }
-                                                                       tabIndex={ tabIndex }
+                                                                       tabIndex={ isFocusWithin() ? tabIndex : 0 }
                                                                >
                                                                        <HStack spacing={ 3 } justify="start">
                                                                                <div className="dataviews-view-list__media-wrapper">
```

</details>

Reply:
I think the issue is that the `<div>` element isn't detected as a focusable.

`TreeGrid` uses the dom package's `focusable` utility internally, and that doesn't seem to be returning the `div` as one of the focusable elements in the row. This is the relevant code:
https://github.com/WordPress/gutenberg/blob/e84b38d63688b2ea9c96a974962d5cee663102ef/packages/components/src/tree-grid/index.tsx#L23-L31

The selector list for `focusable` is here:
https://github.com/WordPress/gutenberg/blob/e84b38d63688b2ea9c96a974962d5cee663102ef/packages/dom/src/focusable.js#L31-L45

Out of interest, why use a `div`?

Reply:
I put together an example of a fix in #57905 for you, but IMO it might be better to use a `<button>` if possible.

I'm also wondering about the suitability of treegrid. I think you should only use it if there's some hierarchy/nesting to the list. Perhaps you have that with parent pages?

It can be fun to style as it makes everything a table 😄

Reply:
> I put together an example of a fix in https://github.com/WordPress/gutenberg/pull/57905 for you

Nice, thanks Dan! Merged it to keep the conversation in one place.

> but IMO it might be better to use a <button> if possible.

I can investigate alternatives.

> I'm also wondering about the suitability of treegrid. I think you should only use it if there's some hierarchy/nesting to the list. Perhaps you have that with parent pages?

Yeah, exactly.

Reply:
> > I'm also wondering about the suitability of treegrid. I think you should only use it if there's some hierarchy/nesting to the list. Perhaps you have that with parent pages?
> 
> Yeah, exactly.

Would using `Composite` be a better alternative? (cc @andrewhayward )